### ISSUE TITLE
The `@stylistic/unit-case` fix is taken from the whole of a word holding a multiplication, so it writes nothing where the word is no dimension and recases what is no unit where it is

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#413]
Found while writing the branch that closes [#297](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/297). Reproduces on `main` before that branch as it does after it.

### What minimal example or steps are needed to reproduce the bug?

A word holding a multiplication, whose parts are not all dimensions. Two shapes, one for each half of the defect.

Nothing is written, and the warning goes away all the same:

```scss
a { b: $VAR*10PX; }
```

More is written than the unit:

```scss
a { b: 10PX*$VAR; }
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "plugins": [
    "@stylistic/stylelint-plugin"
  ],
  "rules": {
    "@stylistic/unit-case": "lower"
  }
}
```

### How did you run Stylelint?

```shell
stylelint "**/*.scss" --customSyntax postcss-scss --config myconfig.json --fix
```

### What did you expect to happen?

`10PX` recased to `10px` in both, and nothing else touched.

### What actually happened?

The first file comes back exactly as it was, and the warning it drew without `--fix` — `1:15 Expected "PX" to be "px"` — is gone from the output, so the run reports nothing and fixes nothing.

The second comes back as `a { b: 10px*$var; }`. The variable is renamed, and every place that reads `$VAR` now points at nothing.

The same two shapes elsewhere:

| code | syntax | `--fix` leaves |
| --- | --- | --- |
| `a { b: #{$a}PX*2REM; }` | `postcss-scss` | unchanged, and the warning about `REM` gone |
| `a { b: @VAR*10PX; }` | `postcss-less` | unchanged, and the warning about `PX` gone |
| `a { b: 1PX*A; }` | css | `a { b: 1px*a; }` — the `A` recased, which is no unit |

### What is the cause?

`lib/rules/unit-case/index.js` reports such a word part by part and writes it whole. The write is decided by `getDimension` reading the whole word as one dimension, and one edit recases the whole of it:

```js
let wordProblem = readMiscasedUnit(valueNode)

if (!wordProblem) return

// …

edits.push({
	start: valueNode.sourceIndex,
	end: valueNode.sourceIndex + value.length,
	text: (primary === `lower` ? dimension.toLowerCase() : dimension.toUpperCase()) + value.slice(dimension.length),
})
```

So the write is both too narrow and too wide. Where the whole word is nothing `getDimension` will read — it opens with a variable, or holds an interpolation — there is no edit at all, and the warning a part drew is dropped by Stylelint as fixed, since the `fix` callback was called. Where it will read it, the edit takes in every character of the word, whether or not any unit stands there.

### Do you have a proposal to fix the bug?

Write part by part, as the report already does. Each miscased unit is a run the reader already measures — `readMiscasedUnit` counts both its ends off the `positions` map `getDimension` gives back — so the same run can carry the text to write, and one edit per unit replaces the one edit per word. Nothing outside a unit is then touched, and a part is written whether or not the word around it reads as a dimension.

**Comment by firefoxic** [comment]
A census of one more shape the first half of this reaches, found on the branch that closes #297.

Where a bang flag stands in a part that is not the last, the text of the edit stops at the flag — `withoutBangFlag(value)` cuts the word there — so every part behind it goes unwritten, while the parts are all reported:

```scss
a { b: 10PX*2REM!default*3EM; }
```

Under `lower`, `postcss-scss`: `PX`, `REM` and `EM` are each reported, and `--fix` leaves `a { b: 10px*2rem!default*3EM; }`. The `EM` warning is dropped as fixed and its unit is never written.

That output is byte for byte what `main` wrote before #297 closed, so nothing about the write moved — what moved is that the `EM` is now named. Before, the second reading named `PX*2REM` and stopped in front of the flag, and the part behind it drew no warning to lose.

The fix this issue asks for — one edit per unit, at the position the report already measures — closes this shape along with the rest.
